### PR TITLE
ONCat Archive Searching

### PR DESCRIPTION
**Pull request by PeterParker**
#### Description of Work

The context of this work is outlined in #22608.  This PR is essentially an implementation of the "New Implementations of `IArchiveSearch`" section of the plan outlined in that issue.

For the next version of Mantid, the old `SNSDataArchive` mechanism (which makes calls to ICAT) will continue to be the default at the SNS, but for HFIR we will now be using `ORNLDataArchive` (which makes calls to ONCat).

I have also made some _very_ slight modifications to existing ICAT functionality elsewhere -- the "Catalog" menu bar item will only be shown if the currently-selected facility has a registered ICAT (so, just ISIS), and `CatalogLogin` only allows facilities with a registered ICAT (again, just ISIS).  The reasoning for this is simply to cut down on confusion about what is and what isn't "the catalog", and the fact that we shouldn't be showing people options that they can't actually use anyway.

#### To Test:

1. Make sure that the existing **SNS** archive-searching functionality is unaffected when using the old (and still the default) `SNSDataArchive` implementation.  After switching on "Data Archive Searching", sensible manual tests (that assume file system access) include things like:

    1. `ws = Load('BSS_83000')`
    1. `ws = Load('PG3_40000.nxs.h5')`
    1. `FileFinder.findRuns('HYS_197151')`

    Confirm ICAT is still being called, just as it always was, by making sure log messages such as the following are still being written out:

   ```
   URL: http://icat.sns.gov:2080/icat-rest-ws/datafile/filename/BSS_83000
   Sending http GET request to: http://icat.sns.gov:2080/icat-rest-ws/datafile/filename/BSS_83000
   Answer from web: 200 OK
   location: /SNS/BSS/IPTS-20464/0/83000/BSS_83000.meta.xml
   location: /SNS/BSS/IPTS-20464/shared/autoreduce/BSS_83000_Xqw.nxs
   location: /SNS/BSS/IPTS-20464/0/83000/preNeXus/BSS_83000_bmon_histo.dat
   location: /SNS/BSS/IPTS-20464/0/83000/preNeXus/BSS_83000_cvinfo.xml
   location: /SNS/BSS/IPTS-20464/0/83000/NeXus/BSS_83000_event.nxs
   location: /SNS/BSS/IPTS-20464/shared/autoreduce/reduction_log/BSS_83000_event.nxs.log
   location: /SNS/BSS/IPTS-20464/0/83000/NeXus/BSS_83000_histo.nxs
   location: /SNS/BSS/IPTS-20464/0/83000/preNeXus/BSS_83000_neutron_event.dat
   location: /SNS/BSS/IPTS-20464/0/83000/preNeXus/BSS_83000_pulseid.dat
   location: /SNS/BSS/IPTS-20464/0/83000/preNeXus/BSS_83000_runinfo.xml
   location: /SNS/BSS/IPTS-20464/shared/autoreduce/BSS_83000_sqw.dat
   location: /SNS/BSS/IPTS-20464/shared/autoreduce/BSS_83000_sqw.nxs
   location: /SNS/BSS/IPTS-20464/shared/autoreduce/BSS_83000_sqw.nxs.png
   found path = /SNS/BSS/IPTS-20464/0/83000/NeXus/BSS_83000_event.nxs
   ```

2. Make sure the new archive-searching functionality works in exactly the same way for the **SNS** as the old functionality does.  Manually change the `Facilities.xml` file to use `ORNLDataSearch` instead of `SNSDataSearch`, and do the same tests above.  It should be clear that ONCat is being called by inspecting the log messages.

3. Make sure the new archive-searching functionality works at **HFIR**.  WAND^2 (HB2C) is the only HFIR instrument for which we've cataloged NeXus "run" files, so the following would be sensible tests:

    1. `ws = Load('HB2C_143210)`
    1. `ws = Load('HB2C_143210.nxs.h5)`
    1. `FileFinder.findRuns('HB2C_143210')`

4. Make sure the ICAT changes are working.  Open up MantidPlot and then:

    1. Select ISIS as your facility.  Make sure the "Catalog" menu bar item is available, and that when you open up the `CatalogLogin` algorithm dialog the default facility is shown as "ISIS" and no other facilities are available.
    1. Select SNS as your facility. Making sure the "Catalog" menu bar item is _not_ available, and that when you open up the `CatalogLogin` algorithm dialog the default facility is also shown as being "ISIS" and no other facilities are available.

5.  Please take note of my CMake changes.  I'm not the best at this, so it's worth giving these a once-over to make sure I've done these in the recommended fashion.  (I've introduced some extra dependencies between sub-projects.)

6. Ensure the release notes are reasonable.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: instrument/Facilities.xml [hunk 1]
```diff
@@ -376,7 +376,11 @@
 </facility>
 
 
-<facility name="HFIR" FileExtensions=".nxs,.dat,.xml">
+<facility name="HFIR" FileExtensions=".nxs,.nxs.h5,.dat,.xml">
```
Comment (peterfpeterson) [inline instrument/Facilities.xml]: Should be
```xml
<facility name="HFIR" delimiter="_" FileExtensions=".nxs,.nxs.h5,.dat,.xml">
```

File: docs/source/release/v3.14.0/framework.rst [hunk 2]
```diff
@@ -29,6 +29,12 @@ Nexus Geometry Loading
...
 
+Archive Searching / ONCat
+-------------------------
+
+- For HFIR instruments that write out raw files with run numbers, we have enabled functionality that allows for the searching of file locations by making calls to ONCat.  To use this, make sure that the "Search Data Archive" option is checked in your "Manage User Directories" settings.  The ``FileFinder`` and algorithms such as :ref:`Load <algm-Load>`  will then accept inputs such as "``HB2C_143210``".
```
Comment (peterfpeterson) [inline docs/source/release/v3.14.0/framework.rst]: My test was showing that `FileFinder.findRuns('HB2C143120')` worked, but with the `_` did not. You may want to list the instruments that are supported.

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
This works well if the file being searched for is in the default facility (using `FileFinder` from python). I believe that is a bug in `FileFinder` rather than this new code.

File: docs/source/release/v3.14.0/framework.rst (hunk 2 above)
Comment (PeterParker) [reply docs/source/release/v3.14.0/framework.rst]: > ... but with the `_` did not.

Is this true even with your suggested fix, above?

**Comment by peterfpeterson** [comment]
Update the `Facilities.xml` and this PR is mostly fine. I ran the following with reasonable results:
```python
prin(str(FileFinder.findRuns('wand143210-143211')))
print(str(FileFinder.findRuns('hb2c143210-143211')))
```
Adding in the `_` in the "hint" doesn't work here, but doesn't for other instruments either. There are a couple of bugs in `FileFinder` that are begging for it to be re-written or at least get a major reworking. Amongst those issues is that it thinks instruments must have no numbers in their names. When they do, it falls back to searching in the default facility and instrument. That **should not** be fixed in this PR.

**Review by peterfpeterson — APPROVED** [review]
Looks good now. The other issues are related to `FileFinder` itself.

**Pull request merged**